src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -486,23 +559,154 @@ InsertValidationResponse insertBufferedRecords(StreamingBuffer streamingBufferTo
+                topicPartition.topic(),
+                sfConnectorConfig.get(
+                    SnowflakeSinkConnectorConfig.VALUE_SCHEMA_REGISTRY_CONFIG_FIELD));
+      }
+
+      extraColumnToType = new HashMap<>();
+      nonNullableColumns = new ArrayList<>();
+
+      for (InsertValidationResponse.InsertError insertError : insertErrors) {
+        Map<String, Object> recordMap;
+        try {
+          recordMap =
+              (Map<String, Object>)
+                  insertedRecordsToBuffer.get((int) insertError.getRowIndex()).value();
+        } catch (Exception e) {
+          recordMap = new HashMap<>();
+        }
+        // For Avro this is a Struct, not a map
+        Map<String, String> tempExtraColumnToType =
+            SchematizationUtils.collectExtraColumnToType(
+                recordMap, insertError.getExtraColNames(), schemaMap);
+        List<String> tempNonNullableColumns = insertError.getMissingNotNullColNames();
+        if (tempNonNullableColumns == null) tempNonNullableColumns = new ArrayList<>();
+        if (tempExtraColumnToType.isEmpty() && tempNonNullableColumns.isEmpty()) {
+          // if not extra column and no non-nullable columns, then add it to response body to be
+          // returned
+          this.insertErrors.add(insertError);
+          this.recordWithError.add(insertedRecordsToBuffer.get((int) insertError.getRowIndex()));
+          // containing other type of error
+        } else {
+          extraColumnToType.putAll(tempExtraColumnToType);
+          nonNullableColumns.addAll(tempNonNullableColumns);
+          failedBuffer.insert(insertedRecordsToBuffer.get((int) insertError.getRowIndex()));
+        }
+      }
+
+      return failedBuffer;
+    }
+
+    /**
+     * From the response collect rows with error and append those with unretryable errors to
+     * recordWithError. Insert the rest into a StreamingBuffer waiting for retry insertion, and
+     * collect the column to be added at the same time. After that the table is altered.
+     *
+     * @param conn the snowflake connection service
+     * @param tableName the name of the table
+     * @param response the response from SDK
+     * @param records the list of records being inserted
+     * @return the buffer containing records to be reinserted
+     */
+    public StreamingBuffer collectRowsWithErrorsAndAlterTable(
+        SnowflakeConnectionService conn,
+        String tableName,
+        InsertValidationResponse response,
+        List<SinkRecord> records) {
+      StreamingBuffer newBufferToInsert =
+          insertRecordsToFailedBuffer(response.getInsertErrors(), records);
+      if (!this.extraColumnToType.isEmpty()) {
+        try {
+          conn.appendColumnsToTable(tableName, this.extraColumnToType);
+        } catch (SnowflakeKafkaConnectorException e) {
+          LOGGER.warn(
+              String.format("[INSERT_BUFFERED_RECORDS] Failure altering table :%s", tableName), e);
```

Comment:
what will we do after we log it? Will we retry?

Reply:
Since this function is called in the `onRetry` clause, yes we will retry.